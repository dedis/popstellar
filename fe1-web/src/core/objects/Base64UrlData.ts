import base64url from 'base64url';

export type Base64UrlDataState = string;

export class Base64UrlData extends String {
  public constructor(value: string) {
    super(value);

    // raise an exception if it's not base64 data
    base64url.decode(value);
  }

  public static encode(text: string | Buffer, encoding: string = 'utf8'): Base64UrlData {
    return new Base64UrlData(this.addPadding(base64url.encode(text, encoding)));
  }

  public decode(): string {
    return base64url.decode(this.valueOf());
  }

  public equals(o: Base64UrlData): boolean {
    return this.toString() === o.toString();
  }

  public toBuffer(): Buffer {
    return base64url.toBuffer(this.valueOf());
  }

  /**
   * Create Base64Url data from Base64 data
   * @param b64 the base64-encoded data
   */
  public static fromBase64(b64: string): Base64UrlData {
    return new Base64UrlData(this.addPadding(base64url.fromBase64(b64)));
  }

  public static fromBuffer(buf: Buffer): Base64UrlData {
    return new Base64UrlData(this.addPadding(base64url.encode(buf)));
  }

  private static addPadding(str: string): string {
    let paddedStr = str;
    while (paddedStr.length % 4 !== 0) {
      paddedStr += '=';
    }
    return paddedStr;
  }

  /**
   * Returns the primitive value used for representing the Base64UrlData,
   * a string
   * If you want to serialize an instance use .toState() instead
   */
  public valueOf(): string {
    return super.valueOf();
  }

  /**
   * Returns the serialized version of the base64url that can for instance be stored
   * in redux stores
   */
  public toState(): Base64UrlDataState {
    return super.valueOf();
  }

  /**
   * Deserializes a previously serializes instance of Base64Url
   */
  public static fromState(base64UrlDataState: Base64UrlDataState): Base64UrlData {
    return new Base64UrlData(base64UrlDataState);
  }
}
