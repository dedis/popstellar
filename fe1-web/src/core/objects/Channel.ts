import { Hash } from './Hash';
import { PublicKey } from './PublicKey';

export type Channel = string;
export const ROOT_CHANNEL: Channel = '/root';

export function getLaoIdFromChannel(ch: Channel): Hash {
  if (!ch || !ch.startsWith(ROOT_CHANNEL)) {
    throw new Error(`Invalid channel path: '${ch}'`);
  }

  const pathComponents = ch.split('/');
  if (pathComponents.length < 3) {
    throw new Error(`LAO is absent from channel path '${ch}'`);
  }

  return new Hash(pathComponents[2]);
}

export function channelFromIds(...args: Hash[]): Channel {
  if (args.length === 0) {
    return ROOT_CHANNEL;
  }

  return `${ROOT_CHANNEL}/${args.map((c) => c.valueOf()).join('/')}`;
}

/**
 * Returns the social channel of the given user.
 *
 * @param laoIdHash - The hash containing the laoID of the currently opened LAO
 * @param userPublicKey - The public key of the user
 */
export function getUserSocialChannel(laoIdHash: Hash, userPublicKey: PublicKey): Channel {
  return `${ROOT_CHANNEL}/${laoIdHash.valueOf()}/social/${userPublicKey}`;
}

/**
 * Returns the coin channel
 *
 * @param laoIdHash - The hash containing the laoID of the currently opened LAO
 */
export function getCoinChannel(laoIdHash: Hash): Channel {
  return `${ROOT_CHANNEL}/${laoIdHash.valueOf()}/coin`;
}

/**
 * Returns the general channel of chirps.
 *
 * @param laoIdHash - The hash containing the laoID of the currently opened LAO
 */
export function getGeneralChirpsChannel(laoIdHash: Hash): Channel {
  return `${ROOT_CHANNEL}/${laoIdHash.valueOf()}/social/chirps`;
}

/**
 * Returns the channel of all the reactions.
 *
 * @param laoIdHash - The hash containing the laoID of the currently opened LAO
 */
export function getReactionChannel(laoIdHash: Hash): Channel {
  return `${ROOT_CHANNEL}/${laoIdHash.valueOf()}/social/reactions`;
}

/**
 * Returns the channel popcha channel for authentication of specific LAO
 * @param laoId - The id of the Lao in which to authenticate
 */
export function getPopchaAuthenticationChannel(laoId: Hash): Channel {
  return `${ROOT_CHANNEL}/${laoId.valueOf()}/authentication`;
}

/** Returns the last part of the channel which is usually an event id
 * Example:
 * Input: /root/laoID/electionID
 * Output: electionID
 *
 * @param channel - The channel whose last component we want to obtain
 *
 * @remarks
 * Using this function is equivalent to making a lot of assumptions about the channel.
 * This can be brittle and using a more validation-heavy logic may be preferable.
 */
export function getLastPartOfChannel(channel: Channel): Hash {
  const channels = channel.split('/');
  return new Hash(channels[channels.length - 1]);
}
