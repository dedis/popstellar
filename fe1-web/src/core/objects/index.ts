export * from './Base64UrlData';
export * from './Channel';
export * from './Hash';
export * from './KeyPair';
export * from './PopToken';
export * from './PrivateKey';
export * from './ProtocolError';
export * from './PublicKey';
export * from './RollCallToken';
export * from './Signature';
export * from './Timestamp';
export * from './WitnessSignature';
