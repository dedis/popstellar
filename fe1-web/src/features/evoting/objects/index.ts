export * from './Election';
export * from './EncryptedVote';
export * from './Question';
export * from './QuestionResult';
export * from './RegisteredVote';
export * from './Vote';
