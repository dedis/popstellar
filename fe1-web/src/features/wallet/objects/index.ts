export * from './Seed';
export { generateToken, getCurrentPopTokenFromStore } from './Token';
export * from './Wallet';
