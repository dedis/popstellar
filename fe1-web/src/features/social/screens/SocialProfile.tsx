import * as React from 'react';
import { useContext, useMemo } from 'react';
import { FlatList, ListRenderItemInfo, Text, View } from 'react-native';
import { useSelector } from 'react-redux';

import { ProfileIcon, TextBlock } from 'core/components';
import ScreenWrapper from 'core/components/ScreenWrapper';
import STRINGS from 'resources/strings';

import { ChirpCard } from '../components';
import { SocialMediaContext } from '../context';
import { SocialHooks } from '../hooks';
import { SocialFeature } from '../interface';
import { Chirp, ChirpState } from '../objects';
import { makeChirpsListOfUser } from '../reducer';
import socialMediaProfileStyles from '../styles/socialMediaProfileStyles';

/**
 * UI for the profile of the current user.
 */

const styles = socialMediaProfileStyles;

const SocialProfile = () => {
  const { currentUserPopTokenPublicKey } = useContext(SocialMediaContext);
  const laoId = SocialHooks.useCurrentLaoId();
  if (!laoId) {
    throw new Error('Impossible to render Social Profile, current lao id is undefined');
  }

  const userChirps = useMemo(
    () => makeChirpsListOfUser(laoId.valueOf())(currentUserPopTokenPublicKey),
    [currentUserPopTokenPublicKey, laoId],
  );
  const userChirpList = useSelector(userChirps);

  if (!currentUserPopTokenPublicKey) {
    return (
      <View style={styles.textUnavailableView}>
        <TextBlock text={STRINGS.social_media_your_profile_unavailable} />
      </View>
    );
  }

  const renderChirpState = ({ item }: ListRenderItemInfo<ChirpState>) => (
    <ChirpCard chirp={Chirp.fromState(item)} currentUserPublicKey={currentUserPopTokenPublicKey} />
  );

  return (
    <ScreenWrapper>
      <View style={styles.viewCenter}>
        <View style={styles.topView}>
          <ProfileIcon publicKey={currentUserPopTokenPublicKey} size={8} scale={10} />
          <View style={styles.textView}>
            <Text style={styles.profileText}>{currentUserPopTokenPublicKey.valueOf()}</Text>
            <Text>{`${userChirpList.length} ${
              userChirpList.length === 1 ? 'chirp' : 'chirps'
            }`}</Text>
          </View>
        </View>
        <View style={styles.userFeed}>
          <FlatList
            data={userChirpList}
            renderItem={renderChirpState}
            keyExtractor={(item) => item.id.toString()}
          />
        </View>
      </View>
    </ScreenWrapper>
  );
};

export default SocialProfile;

export const SocialProfileScreen: SocialFeature.SocialScreen = {
  id: STRINGS.social_media_navigation_tab_profile,
  Component: SocialProfile,
};
