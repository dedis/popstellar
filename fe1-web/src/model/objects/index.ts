import * as w from './wallet';

export const Wallet = w;

export * from './Base64Url';
export * from './Channel';
export * from './ConnectToLao';
export * from './Chirp';
export * from './Election';
export * from './Hash';
export * from './KeyPair';
export * from './Lao';
export * from './LaoEvent';
export * from './LaoEventBuilder';
export * from './Meeting';
export * from './PopToken';
export * from './PrivateKey';
export * from './PublicKey';
export * from './Reaction';
export * from './RollCall';
export * from './Signature';
export * from './Timestamp';
export * from './WitnessSignature';
