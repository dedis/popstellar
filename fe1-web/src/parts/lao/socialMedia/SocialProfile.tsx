import * as React from 'react';
import PropTypes from 'prop-types';
import { Chirp, ChirpState, PublicKey } from 'model/objects';
import TextBlock from 'components/TextBlock';
import ChirpCard from 'components/ChirpCard';
import ProfileIcon from 'components/ProfileIcon';
import { FlatList, ListRenderItemInfo, Text, View } from 'react-native';
import STRINGS from 'res/strings';
import { makeChirpsListOfUser } from 'store';
import { useSelector } from 'react-redux';
import socialMediaProfileStyles from 'styles/stylesheets/socialMediaProfileStyles';

/**
 * UI for the profile of the current user.
 */

const styles = socialMediaProfileStyles;

const SocialProfile = (props: IPropTypes) => {
  const { currentUserPublicKey } = props;
  const userChirps = makeChirpsListOfUser(currentUserPublicKey);
  const userChirpList = useSelector(userChirps);

  if (!currentUserPublicKey || currentUserPublicKey.valueOf() === '') {
    return (
      <View style={styles.textUnavailableView}>
        <TextBlock text={STRINGS.social_media_your_profile_unavailable} />
      </View>
    );
  }

  const renderChirpState = ({ item }: ListRenderItemInfo<ChirpState>) => (
    <ChirpCard chirp={Chirp.fromState(item)} currentUserPublicKey={currentUserPublicKey} />
  );

  return (
    <View style={styles.viewCenter}>
      <View style={styles.topView}>
        <ProfileIcon publicKey={currentUserPublicKey} size={8} scale={10} />
        <View style={styles.textView}>
          <Text style={styles.profileText}>{currentUserPublicKey.valueOf()}</Text>
          <Text>{`${userChirpList.length} ${
            userChirpList.length === 1 ? 'chirp' : 'chirps'
          }`}</Text>
        </View>
      </View>
      <View style={styles.userFeed}>
        <FlatList
          data={userChirpList}
          renderItem={renderChirpState}
          keyExtractor={(item) => item.id.toString()}
        />
      </View>
    </View>
  );
};

const propTypes = {
  currentUserPublicKey: PropTypes.instanceOf(PublicKey).isRequired,
};

SocialProfile.prototype = propTypes;

type IPropTypes = {
  currentUserPublicKey: PublicKey;
};

export default SocialProfile;
