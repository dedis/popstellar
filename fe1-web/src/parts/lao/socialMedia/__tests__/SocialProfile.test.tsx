import React from 'react';
import { render } from '@testing-library/react-native';
import { PublicKey } from 'model/objects';
import SocialProfile from '../SocialProfile';

const mockEmptyPublicKey = new PublicKey('');

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn().mockImplementation(() => []),
}));

describe('SocialProfile', () => {
  it('renders correctly with an empty user public key', () => {
    const component = render(<SocialProfile currentUserPublicKey={mockEmptyPublicKey} />).toJSON();
    expect(component).toMatchSnapshot();
  });
});
