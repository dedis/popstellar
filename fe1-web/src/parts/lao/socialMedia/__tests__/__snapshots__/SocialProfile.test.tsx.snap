// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SocialProfile renders correctly with an empty user public key 1`] = `
<View
  style={
    Object {
      "alignSelf": "center",
      "marginTop": 20,
      "width": 600,
    }
  }
>
  <Text
    style={
      Object {
        "color": "black",
        "fontSize": 25,
        "marginHorizontal": 10,
        "textAlign": "center",
      }
    }
  >
    You do not have a social media profile yet, be sure to have participated in a roll call.
  </Text>
</View>
`;
